File: tail_recursion.ex
# Calculate the nth Fibonacci number, f(n). Using invariants
defmodule Fibo do
  def of(n) when n < 1, do: IO.puts "n should be positive"
  def of(n), do: of(n, 1, 1) 
  def of(1, _, result), do: result
  def of(n, next, result), do: of(n - 1, next + result, next)
end
 
# Display the Fibonacci sequence.
Enum.each 1..5000, &IO.puts "fibo(#{&1}) = #{Fibo.of(&1)}"

File: tail_recursion.rb
# Calculate the nth Fibonacci number, f(n). Using invariants
def fibo_tr(n, next_item, result)
  if n == 0
    0
  elsif n < 2
    result
  else
    return fibo_tr(n - 1, result, result + next_item)
  end
end

def fibo (n)
  fibo_tr(n, 0, 1)
end 

# Display the Fibonacci sequence.
(1..5000).each do |number|
  puts "fibo(#{number}) = #{fibo(number)}"
end